hammer/hammer.go

```diff
@@ -0,0 +1,362 @@
+// Copyright 2024 The Tessera authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// hammer is a tool to load test a Tessera log.
+package main
+
+import (
+	"context"
+	crand "crypto/rand"
+	"errors"
+	"flag"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gdamore/tcell/v2"
+	"github.com/rivo/tview"
+	"github.com/transparency-dev/trillian-tessera/client"
+	"golang.org/x/mod/sumdb/note"
+	"k8s.io/klog/v2"
+)
+
+func init() {
+	flag.Var(&logURL, "log_url", "Log storage root URL (can be specified multiple times), e.g. https://log.server/and/path/")
+}
+
+var (
+	logURL multiStringFlag
+
+	logPubKey = flag.String("log_public_key", os.Getenv("TILES_LOG_PUBLIC_KEY"), "Public key for the log. This is defaulted to the environment variable TILES_LOG_PUBLIC_KEY")
+
+	maxReadOpsPerSecond = flag.Int("max_read_ops", 20, "The maximum number of read operations per second")
+	numReadersRandom    = flag.Int("num_readers_random", 4, "The number of readers looking for random leaves")
+	numReadersFull      = flag.Int("num_readers_full", 4, "The number of readers downloading the whole log")
+
+	maxWriteOpsPerSecond = flag.Int("max_write_ops", 0, "The maximum number of write operations per second")
+	numWriters           = flag.Int("num_writers", 0, "The number of independent write tasks to run")
+
+	leafMinSize = flag.Int("leaf_min_size", 0, "Minimum size in bytes of individual leaves")
+
+	showUI = flag.Bool("show_ui", true, "Set to false to disable the text-based UI")
+
+	hc = &http.Client{
+		Transport: &http.Transport{
+			MaxIdleConns:        256,
+			MaxIdleConnsPerHost: 256,
+			DisableKeepAlives:   false,
+		},
+		Timeout: 5 * time.Second,
+	}
+)
+
+func main() {
+	klog.InitFlags(nil)
+	flag.Parse()
+
+	ctx := context.Background()
+
+	logSigV, err := note.NewVerifier(*logPubKey)
+	if err != nil {
+		klog.Exitf("failed to create verifier: %v", err)
+	}
+
+	f, w := newLogClientsFromFlags()
+
+	var cpRaw []byte
```

Comment:
There's a catch-22 here at the mo' - Tessera logs don't create a checkpoint until they've added an entry, so the hammer bails here before it can add any entries.
Perhaps the right answer is to have the logs create a zero checkpoint, but it's also sorta special-case there too.

Reply:
We discussed this some weeks ago and the conclusion is that logs ought to create a checkpoint for the 0 size. There was some notion being floated that this was out of bounds from the specs, but it's explicitly in the spec:

https://github.com/C2SP/C2SP/blob/main/tlog-checkpoint.md : `The second line is the tree size, the ASCII decimal representation of the number of leaves in the tree, with no leading zeroes (unless the tree is empty, in which case the tree size is 0).`